diff --git a/src/analyze/filters.rs b/src/analyze/filters.rs
index 8264b1f..b7b6194 100644
--- a/src/analyze/filters.rs
+++ b/src/analyze/filters.rs
@@ -1,21 +1,16 @@
 use crate::analyze::{LanguageWithStopwords, stopwords, u17_to_lower::unicode_v17_char_to_lower};
 
+// The maximum byte length of a character, after lowercasing.
+// Verified with `test_max_unicode_lowercased_length`.
+const MAX_LOWERCASED_BYTE_LENGTH: usize = 4;
+
 pub(crate) fn lowercase_chars_in_place(s: &mut String) {
     let original_byte_length = s.len();
 
-    // Compute the number of extra bytes we need to store the lowercased
-    // version of the string. Since lowercasing can result in multiple characters,
-    // it's difficult to pick an easy upper bound, so we'll just compute an
-    // exact one.
-    let required_bytes_for_lowercase = s
-        .chars()
-        .flat_map(|c| unicode_v17_char_to_lower(c))
-        .map(|c| c.len_utf8())
-        .sum::<usize>();
-
-    // Pad the string with enough space for the full lowercased output. Lowercasing
-    // can either expand or shrink in bytes, depending on the character.
-    s.extend(std::iter::repeat('\0').take(required_bytes_for_lowercase));
+    // Pad the string with enough space for the full lowercased output.
+    // Lowercasing will, at most, expand each character to `MAX_LOWERCASED_BYTE_LENGTH` bytes.
+    let additional_scratch = s.chars().count() * MAX_LOWERCASED_BYTE_LENGTH;
+    s.extend(std::iter::repeat('\0').take(additional_scratch));
 
     // Split the string into two parts, the original and lowercased part.
     // We'll read from the original part and write to the lowercased part.
@@ -33,7 +28,10 @@ pub(crate) fn lowercase_chars_in_place(s: &mut String) {
                 dst_idx += lc.encode_utf8(&mut lowercased_bytes[dst_idx..]).len();
             }
         }
-        debug_assert_eq!(dst_idx, required_bytes_for_lowercase);
+        assert!(
+            dst_idx <= additional_scratch,
+            "lowercased output should not exceed reserved space"
+        );
     }
 
     // Truncate the extra bytes we didn't use, then drain the original part.
@@ -1556,3 +1554,23 @@ fn fold_non_ascii_char(c: char) -> Option<&'static str> {
         _ => None
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::analyze::{
+        filters::MAX_LOWERCASED_BYTE_LENGTH, u17_to_lower::unicode_v17_char_to_lower,
+    };
+
+    #[test]
+    fn test_max_unicode_lowercased_length() {
+        let mut max_lowercased_byte_len = 0;
+        for c in 0..=char::MAX as u32 {
+            if let Some(c) = char::from_u32(c) {
+                let lowercased_byte_length =
+                    unicode_v17_char_to_lower(c).map(|lc| lc.len_utf8()).sum();
+                max_lowercased_byte_len = max_lowercased_byte_len.max(lowercased_byte_length);
+            }
+        }
+        assert_eq!(max_lowercased_byte_len, MAX_LOWERCASED_BYTE_LENGTH);
+    }
+}
